File: libs/src/python/salmon-fs/utils/salmon_file_exporter.py
#!/usr/bin/env python3
'''
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
'''
from __future__ import annotations

import concurrent
import math
import time
from concurrent.futures import ThreadPoolExecutor, Future, ProcessPoolExecutor
from multiprocessing import shared_memory
from multiprocessing.shared_memory import SharedMemory
from typing import Any, Callable

from typeguard import typechecked

from convert.bit_converter import BitConverter
from file.ireal_file import IRealFile
from iostream.random_access_stream import RandomAccessStream
from salmon.iostream.salmon_stream import SalmonStream
from salmonfs.salmon_file import SalmonFile
from utils.salmon_file_utils import SalmonFileUtils


@typechecked
def export_file(index: int, final_part_size: int, final_running_threads: int, file_size: int,
                real_file_to_export: IRealFile,
                real_file: IRealFile,
                shm_total_bytes_read_name: str,
                shm_cancel_name: str,
                buffer_size: int, key: bytearray, integrity: bool, hash_key: bytearray | None, chunk_size: int,
                enable_log_details: bool = False):
    file_to_export: SalmonFile = SalmonFile(real_file_to_export)
    file_to_export.set_allow_overwrite(True)
    file_to_export.set_encryption_key(key)
    file_to_export.set_apply_integrity(integrity, hash_key, chunk_size)

    shm_total_bytes_read = SharedMemory(shm_total_bytes_read_name)
    total_bytes_written = memoryview(shm_total_bytes_read.buf[index * 8:(index + 1) * 8])

    start: int = final_part_size * index
    length: int
    if index == final_running_threads - 1:
        length = file_size - start
    else:
        length = final_part_size
    try:
        export_file_part(file_to_export, real_file, start, length, total_bytes_written,
                         buffer_size,
                         shm_cancel_name, enable_log_details)
    except Exception as ex:
        raise ex
    finally:
        if shm_total_bytes_read is not None:
            total_bytes_written.release()
            shm_total_bytes_read.close()
            shm_total_bytes_read.unlink()


@typechecked
def export_file_part(file_to_export: SalmonFile, real_file: IRealFile, start: int, count: int,
                     total_bytes_written: memoryview, buffer_size: int,
                     shm_cancel_name: str | None = None, enable_log_details: bool = False):
    """
     * Export a file part from the drive.
     *
     * @param file_to_export      The file the part belongs to
     * @param export_file        The file to copy the exported part to
     * @param start             The start position on the file
     * @param count             The length of the bytes to be decrypted
     * @param total_bytes_written The total bytes that were written to the external file
    """

    shm_cancel_data: memoryview | None = None
    shm_cancel: SharedMemory | None = None
    if shm_cancel_name is not None:
        shm_cancel = SharedMemory(shm_cancel_name, size=1)
        shm_cancel_data = shm_cancel.buf

    start_time: int = int(time.time() * 1000)
    total_part_bytes_written: int = 0

    target_stream: RandomAccessStream | None = None
    source_stream: RandomAccessStream | None = None

    try:
        target_stream = real_file.get_output_stream()
        target_stream.set_position(start)

        source_stream = file_to_export.get_input_stream()
        source_stream.set_position(start)

        v_bytes: bytearray = bytearray(buffer_size)
        bytes_read: int
        if enable_log_details:
            print("SalmonFileExporter: FilePart: " + file_to_export.get_real_file().get_base_name() + ": "
                  + file_to_export.get_base_name() + " start = " + str(start) + " count = " + str(count))

        while (bytes_read := source_stream.read(v_bytes, 0,
                                                min(len(v_bytes), count - total_part_bytes_written))) > 0 \
                and total_part_bytes_written < count:
            if shm_cancel_data is not None and shm_cancel_data[0]:
                break
            target_stream.write(v_bytes, 0, bytes_read)
            total_part_bytes_written += bytes_read
            total_bytes_written[:8] = BitConverter.to_bytes(total_part_bytes_written, 8)[0:8]

        if enable_log_details:
            total: int = int(time.time() * 1000) - start_time
            print("SalmonFileExporter: File Part: " + file_to_export.get_base_name() + " exported " + str(
                total_part_bytes_written)
                  + " bytes in: " + str(total) + " ms"
                  + ", avg speed: " + str(total_bytes_written[0] / float(total)) + " bytes/sec")

    except Exception as ex:
        print(ex)
        raise ex
    finally:
        if target_stream is not None:
            target_stream.flush()
            target_stream.close()

        if source_stream is not None:
            source_stream.close()

        if shm_cancel is not None:
            shm_cancel.close()
            shm_cancel.unlink()


@typechecked
class SalmonFileExporter:
    __DEFAULT_BUFFER_SIZE = 512 * 1024
    """
     * The global default buffer size to use when reading/writing on the SalmonStream.
    """

    __DEFAULT_THREADS = 1
    """
     * The global default threads to use.
    """

    __MIN_FILE_SIZE = 2 * 1024 * 1024
    """
     * Minimum file size to use parallelism. Anything less will use single thread.
    """

    __enableMultiThread: bool = True
    """
     * True if multithreading is enabled.
    """

    __enableLog: bool = False
    __enableLogDetails: bool = False

    def __init__(self, buffer_size: int, threads: int, multi_cpu: bool = False):
        """
         * Constructs a file exporter that can be used to export files from the drive
         *
         * @param buffer_size Buffer size to be used when decrypting files.
         *                   If using integrity self value has to be a multiple of the Chunk size.
         *                   If not using integrity it should be a multiple of the AES block size for better performance
         * @param threads
         * :multi_cpu:  Utilize multiple cpus. Windows does not have a fast fork() so it has a very slow startup
        """

        self.__buffer_size: int = 0
        """
         * Current buffer size.
        """

        self.__threads: int = 0
        """
         * Current threads.
        """

        self.__stopped: bool = True
        """
         * True if last job was stopped by the user.
        """

        self.__shm_cancel = shared_memory.SharedMemory(create=True, size=1)
        """
        Shared memory to notify for task cancellations 
        """

        self.__failed: bool = False
        """
         * Failed if last job was failed.
        """

        self.__lastException: Exception | None = None
        """
         * Last exception occurred.
        """

        self.__executor: ThreadPoolExecutor | ProcessPoolExecutor | None = None
        """
         * The executor to be used for running parallel exports.
        """

        if buffer_size == 0:
            buffer_size = SalmonFileExporter.__DEFAULT_BUFFER_SIZE
        if threads == 0:
            threads = SalmonFileExporter.__DEFAULT_THREADS
        self.__buffer_size = buffer_size
        self.__threads = threads

        self.__executor = ThreadPoolExecutor(self.__threads) if not multi_cpu else ProcessPoolExecutor(self.__threads)

    @staticmethod
    def set_enable_log(value: bool):
        SalmonFileExporter.__enableLog = value

    @staticmethod
    def set_enable_log_details(value: bool):
        SalmonFileExporter.__enableLogDetails = value

    def is_running(self) -> bool:
        return not self.__stopped

    def stop(self):
        self.__stopped = True

    def export_file(self, file_to_export: SalmonFile, export_dir: IRealFile, filename: str | None,
                    delete_source: bool,
                    integrity: bool, on_progress: Callable[[int, int], Any] | None) -> IRealFile | None:
        """
         * Export a file from the drive to the external directory path
         *
         * @param file_to_export The file that will be exported
         * @param export_dir    The external directory the file will be exported to
         * @param filename     The filename to use
         * @param delete_source Delete the source file when the export finishes successfully
         * @param integrity    True to verify integrity
        """

        if self.is_running():
            Exception("Another export is running")
        if file_to_export.is_directory():
            raise Exception("Cannot export directory, use SalmonFileCommander instead")

        exported_file: IRealFile | None = None
        filename = filename if filename is not None else file_to_export.get_base_name()
        try:
            if not self.__enableMultiThread and self.__threads != 1:
                raise NotImplementedError("Multithreading is not supported")

            start_time: int = 0
            self.__stopped = False
            if SalmonFileExporter.__enableLog:
                start_time = int(time.time() * 1000)

            total_bytes_written = [0]
            self.__failed = False

            if not export_dir.exists():
                export_dir.mkdir()
            exported_file = export_dir.create_file(filename)
            # we use the drive hash key for integrity verification
            file_to_export.set_verify_integrity(integrity, None)

            file_size: int = file_to_export.get_size()
            part_size: int = file_size
            running_threads: int

            # make sure we allocate enough space for the file
            target_stream: RandomAccessStream = exported_file.get_output_stream()
            target_stream.set_length(file_size)
            target_stream.close()

            if file_size > SalmonFileExporter.__MIN_FILE_SIZE:
                part_size = int(math.ceil(file_size / float(self.__threads)))

                # if we want to check integrity we align to the chunk size otherwise to the AES Block
                min_part_size: int = SalmonFileUtils.get_minimum_part_size(file_to_export)

                # calculate the last part size
                rem: int = part_size % min_part_size
                if rem != 0:
                    part_size += min_part_size - rem

                running_threads = int(math.ceil(file_size / float(part_size)))
            else:
                running_threads = 1

            if running_threads == 0:
                running_threads = 1

            final_part_size: int = part_size
            final_running_threads: int = running_threads

            shm_total_bytes_read = shared_memory.SharedMemory(create=True, size=8 * final_running_threads)
            shm_total_bytes_read_name = shm_total_bytes_read.name

            shm_cancel_name = self.__shm_cancel.name
            self.__shm_cancel.buf[0] = 0

            fs: list[Future] = []
            for i in range(0, running_threads):
                fs.append(self.__executor.submit(export_file, i, final_part_size,
                                                 final_running_threads,
                                                 file_size,
                                                 file_to_export.get_real_file(),
                                                 exported_file,
                                                 shm_total_bytes_read_name,
                                                 shm_cancel_name,
                                                 # on_progress,
                                                 self.__buffer_size,
                                                 file_to_export.get_encryption_key(),
                                                 integrity,
                                                 file_to_export.get_hash_key(),
                                                 file_to_export.get_requested_chunk_size(),
                                                 SalmonFileExporter.__enableLogDetails))
            total_bytes_read = 0
            while True:
                n_total_bytes_read = 0
                for i in range(0, len(fs)):
                    chunk_bytes_read = bytearray(shm_total_bytes_read.buf[i * 8:(i + 1) * 8])
                    n_total_bytes_read += BitConverter.to_long(chunk_bytes_read, 0, 8)

                if total_bytes_read != n_total_bytes_read:
                    total_bytes_read = n_total_bytes_read
                    on_progress(total_bytes_read, file_to_export.get_size())

                complete: bool = True
                for f in fs:
                    complete &= f.done()
                if complete:
                    break
                time.sleep(0.25)

            for f in concurrent.futures.as_completed(fs):
                try:
                    # catch any errors within the children processes
                    f.result()
                except Exception as ex1:
                    print(ex1)
                    self.__lastException = ex1
                    self.__failed = True
                    # cancel all tasks
                    self.stop()

            shm_total_bytes_read.close()
            shm_total_bytes_read.unlink()

            if self.__stopped:
                exported_file.delete()
            elif delete_source:
                file_to_export.get_real_file().delete()
            if self.__lastException is not None:
                raise self.__lastException
            if SalmonFileExporter.__enableLog:
                total: int = int(time.time() * 1000) - start_time
                print("SalmonFileExporter AesType: " + SalmonStream.get_aes_provider_type().name
                      + " File: " + file_to_export.get_base_name() + " verified and exported "
                      + str(total_bytes_written[0]) + " bytes in: " + str(total) + " ms"
                      + ", avg speed: " + str(total_bytes_written[0] / float(total)) + " bytes/sec")

        except Exception as ex:
            print(ex)
            self.__failed = True
            self.__stopped = True
            raise ex

        if self.__stopped or self.__failed:
            self.__stopped = True
            return None

        self.__stopped = True
        return exported_file

    def _finalize(self):
        self.close()

    def close(self):
        self.__executor.shutdown(False)
        self.__shm_cancel.close()
        self.__shm_cancel.unlink()

File: libs/src/typescript/salmon-fs/utils/salmon_file_exporter.ts
/*
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

import { IRealFile } from "../file/ireal_file.js";
import { SalmonFile } from "../salmonfs/salmon_file.js";
import { SalmonFileUtils } from "../utils/salmon_file_utils.js";
import { SalmonStream } from "../../salmon-core/salmon/io/salmon_stream.js";
import { exportFilePart } from "./salmon_file_exporter_helper.js";
import { ProviderType } from "../../salmon-core/salmon/io/provider_type.js";
import { RandomAccessStream } from "../../salmon-core/io/random_access_stream.js";
import { SalmonIntegrity } from "../../salmon-core/salmon/integrity/salmon_integrity.js";
import { SalmonIntegrityException } from "../../salmon-core/salmon/integrity/salmon_integrity_exception.js";
import { IOException } from "../../salmon-core/io/io_exception.js";
import { SalmonAuthException } from "../salmonfs/salmon_auth_exception.js";

export class SalmonFileExporter {
    static #workerPath = './lib/salmon-fs/utils/salmon_file_exporter_worker.js';

    /**
     * The global default buffer size to use when reading/writing on the SalmonStream.
     */
    private static readonly DEFAULT_BUFFER_SIZE: number = 512 * 1024;

    /**
     * The global default threads to use.
     */
    static readonly #DEFAULT_THREADS: number = 1;

    /**
     * True if multithreading is enabled.
     */
    static readonly #enableMultiThread: boolean = true;

    static #enableLog: boolean = false;
    static #enableLogDetails: boolean = false;

    /**
     * Current buffer size.
     */
    readonly #bufferSize: number = 0;

    /**
     * Current threads.
     */
    readonly #threads: number = 1;

    /**
     * True if last job was stopped by the user.
     */
    #stopped: boolean[] = [true];

    /**
     * Failed if last job was failed.
     */
    #failed: boolean = false;

    /**
     * Last exception occurred.
     */
    #lastException: Error | unknown | null = null;

    #promises: Promise<any>[] = [];

    #workers: any[] = [];

    public constructor(bufferSize: number, threads: number) {
        if (bufferSize == 0)
            bufferSize = SalmonFileExporter.DEFAULT_BUFFER_SIZE;
        if (threads == 0)
            threads = SalmonFileExporter.#DEFAULT_THREADS;
        this.#bufferSize = bufferSize;
        this.#threads = threads;
        if (typeof process !== 'object') {
            // multiple writers in the browser use crswap files that overwrite 
            // each other so falling back to 1 thread
            this.#threads = 1;
        }
    }

    public static setEnableLog(value: boolean): void {
        SalmonFileExporter.#enableLog = value;
    }
    public static getEnableLog(): boolean {
        return SalmonFileExporter.#enableLog;
    }

    public static setEnableLogDetails(value: boolean): void {
        SalmonFileExporter.#enableLogDetails = value;
    }
    public static getEnableLogDetails(): boolean {
        return SalmonFileExporter.#enableLogDetails;
    }

    public isRunning(): boolean {
        return !this.#stopped[0];
    }

    /**
     *
     */
    public stop(): void {
        this.#stopped[0] = true;
        let msg = {message: 'stop'};
        for(let i=0; i<this.#workers.length; i++)
            this.#workers[i].postMessage(msg);
    }

    /**
     * Export a file from the drive to the external directory path
     *
     * @param fileToExport The file that will be exported
     * @param exportDir    The external directory the file will be exported to
     * @param filename     The filename to use
     * @param deleteSource Delete the source file when the export finishes successfully
     * @param integrity    True to verify integrity
     */
    public async exportFile(fileToExport: SalmonFile, exportDir: IRealFile, filename: string,
        deleteSource: boolean, integrity: boolean, onProgress: ((position: number, length: number) => void | null)): Promise<IRealFile | null> {
        if (this.isRunning())
            throw new Error("Another export is running");
        if (await fileToExport.isDirectory())
            throw new Error("Cannot export directory, use SalmonFileCommander instead");

        let exportFile: IRealFile;
        filename = filename != null ? filename : await fileToExport.getBaseName();
        try {
            if (!SalmonFileExporter.#enableMultiThread && this.#threads != 1)
                throw new Error("Multithreading is not supported");

            let startTime: number = 0;
            this.#stopped[0] = false;
            this.#lastException = null;
            if (SalmonFileExporter.#enableLog) {
                startTime = Date.now();
            }
            let totalBytesWritten: number[] = [0];
            this.#failed = false;

            if (!await exportDir.exists())
                await exportDir.mkdir();
            exportFile = await exportDir.createFile(filename);
            // we use the drive hash key for integrity verification
            await fileToExport.setVerifyIntegrity(integrity, null);

            let fileSize: number = await fileToExport.getSize();
            let runningThreads: number = 1;
            let partSize: number = fileSize;

            // for python we make sure to allocate enough space for the file
            let targetStream: RandomAccessStream = await exportFile.getOutputStream();
            await targetStream.setLength(fileSize);
            await targetStream.close();

            // if we want to check integrity we align to the chunk size otherwise to the AES Block
            let minPartSize: number = await SalmonFileUtils.getMinimumPartSize(fileToExport);
            if (partSize > minPartSize && this.#threads > 1) {
                partSize = Math.ceil(fileSize / this.#threads);
                if (partSize > minPartSize)
                    partSize -= partSize % minPartSize;
                else
                    partSize = minPartSize;
                runningThreads = Math.floor(fileSize / partSize);
            }

            if(runningThreads == 1) {
                await exportFilePart(fileToExport, exportFile, 0, fileSize, totalBytesWritten, onProgress, this.#bufferSize, this.#stopped, SalmonFileExporter.#enableLogDetails);
            } else {
                await this.#submitExportJobs(runningThreads, partSize, fileToExport, exportFile, totalBytesWritten, integrity, onProgress);
            }

            if (this.#stopped[0])
                await exportFile.delete();
            else if (deleteSource)
                await fileToExport.getRealFile().delete();
            if (this.#lastException != null)
                throw this.#lastException;
            if (SalmonFileExporter.#enableLog && !this.#failed && !this.#stopped[0]) {
                let total: number = Date.now() - startTime;
                console.log("SalmonFileExporter AesType: " + ProviderType[SalmonStream.getAesProviderType()]
                    + " File: " + await fileToExport.getBaseName() + " verified and exported "
                    + totalBytesWritten[0] + " bytes in: " + total + " ms"
                    + ", avg speed: " + totalBytesWritten[0] / total + " Kbytes/sec");
            }
        } catch (ex) {
            console.error(ex);
            this.#failed = true;
            this.#stopped[0] = true;
            this.#lastException = ex;
            throw ex;
        }

        if (this.#stopped[0] || this.#failed) {
            this.#stopped[0] = true;
            return null;
        }
        this.#stopped[0] = true;
        return exportFile;
    }

    async #submitExportJobs(runningThreads: number, partSize: number, fileToExport: SalmonFile, exportedFile: IRealFile, 
        totalBytesWritten: number[], integrity: boolean, onProgress: ((position: number, length: number)=>void) | null) {
            let fileSize: number = await fileToExport.getSize();
            let bytesWritten: number[] = new Array(runningThreads);
            bytesWritten.fill(0);
            this.#promises = [];
            for (let i = 0; i < runningThreads; i++) {
                this.#promises.push(new Promise(async (resolve, reject) => {
                    let fileToExportHandle: any = await fileToExport.getRealFile().getPath();
                    let exportedFileHandle: any = await exportedFile.getPath();
                    if (typeof process !== 'object') {
                        if(this.#workers[i] == null)
                            this.#workers[i] = new Worker(SalmonFileExporter.#workerPath, { type: 'module' });
                        this.#workers[i].addEventListener('message', (event: any) => {
                            if(event.data.message == 'progress' && onProgress != null) {
                                bytesWritten[event.data.index] = event.data.position;
                                totalBytesWritten[0] = 0;
                                for (let i = 0; i < bytesWritten.length; i++) {
                                    totalBytesWritten[0] += bytesWritten[i];
                                }
                                onProgress(totalBytesWritten[0], fileSize);
                            }
                            else if(event.data.message == 'complete') {
                                resolve(event.data);
                            } else if(event.data.message == 'error') {
                                reject(event.data);
                            }
                        });
                        this.#workers[i].addEventListener('error', (event: any) => {
                            reject(event);
                        });
                    } else {
                        const { Worker } = await import("worker_threads");
                        if(this.#workers[i] == null)
                            this.#workers[i] = new Worker(SalmonFileExporter.#workerPath);
                        this.#workers[i].on('message', (event: any) => {
                            if(event.message == 'progress' && onProgress != null) {
                                bytesWritten[event.index] = event.position;
                                totalBytesWritten[0] = 0;
                                for (let i = 0; i < bytesWritten.length; i++) {
                                    totalBytesWritten[0] += bytesWritten[i];
                                }
                                onProgress(totalBytesWritten[0], fileSize);
                            }
                            else if(event.message == 'complete') {
                                resolve(event);
                            } else if(event.message == 'error') {
                                reject(event);
                            }
                        });
                        this.#workers[i].on('error', (event: any) => {
                            reject(event);
                        });
                    }
    
                    let start: number = partSize * i;
                    let length: number;
                    if (i == runningThreads - 1)
                        length = fileSize - start;
                    else
                        length = partSize;
                    this.#workers[i].postMessage({message: 'start', 
                        index: i, 
                        fileToExportHandle: fileToExportHandle, 
                        exportFileClassType: fileToExport.getRealFile().constructor.name,
                        start: start, length: length, 
                        exportedFileHandle: exportedFileHandle,
                        exportedFileClassType: exportedFile.constructor.name,
                        key: fileToExport.getEncryptionKey(),
                        integrity: integrity,
                        hash_key: fileToExport.getHashKey(), 
                        chunk_size: fileToExport.getRequestedChunkSize(),
                        bufferSize: this.#bufferSize,
                        enableLogDetails: SalmonFileExporter.#enableLogDetails
                    });
                }));
            }
            await Promise.all(this.#promises).then((results: any) => {
                totalBytesWritten[0] = 0;
                for (let i = 0; i < results.length; i++) {
                    totalBytesWritten[0] += results[i].totalBytesWritten;
                }
            }).catch((err) => {
                // deserialize the error
                if (err.error != undefined ) {
                    if(err.type == 'SalmonIntegrityException')
                        err = new SalmonIntegrityException(err.error);
                    else if(err.type == 'SalmonAuthException')
                        err = new SalmonAuthException(err.error);
                    else
                        err = new Error(err.error);
                }
                console.error(err);
                this.#failed = true;
                this.#lastException = err;
                this.stop();
                throw new IOException("Error during export", err);
            });
    }

    public close(): void {
        for(let i=0; i<this.#workers.length; i++) {
            this.#workers[i].terminate();
            this.#workers[i] = null;       
        }
        this.#promises = [];
    }

    public static setWorkerPath(path: string) {
        SalmonFileExporter.#workerPath = path;
    }

    public static getWorkerPath(): string {
        return SalmonFileExporter.#workerPath;
    }
}
